src/ucp/wireup/wireup_cm.c

```diff
@@ -378,13 +378,16 @@ static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
  */
 static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
 {
-    ucp_ep_h ucp_ep = req->send.ep;
+    ucp_ep_h ucp_ep            = req->send.ep;
+    /* the EP can be closed/destroyed from err callback */
+    ucs_async_context_t *async = &ucp_ep->worker->async;
 
-    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    UCS_ASYNC_BLOCK(async);
     ucp_ep_cm_disconnect_cm_lane(ucp_ep);
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_CALLBACK));
     ucp_request_put(req);
-    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+    ucp_ep_invoke_err_cb(ucp_ep, UCS_ERR_CONNECTION_RESET);
```

Comment:
why is the err cb invoked any way? are we getting here only in case of remote disconnect?

Reply:
yes, `ucp_ep_cm_disconnect_flushed_cb` is a final callback of close protocol started from `ucp_cm_disconnect_cb`

Reply:
but why is this an error? in case of destroying a connection after finishing communication, it's a valid flow.

Reply:
We notify user about remote disconnect by err callback (if set, all required checks inside `ucp_ep_invoke_err_cb`) with status UCS_ERR_CONNECTION_RESET

Reply:
dont really see why it should be an error case. but on the other hand, how can we tell the reason why the client disconnected..? if it's sudden disconnect, and then it's an error, or after finishing communication, and then it's not really an error.

Reply:
remote disconnect it's not an error case, but it should be handled correspondingly by user. Err callback is a mechanism which we use to delivery this type of event, just like ECONNRESET on socket fd

Reply:
yeah but we deliver it as an error to the user, when it's really not an error - it's a valid remote disconnect.
but ok, if tcp does the same, then we're ok.

Reply:
i though we need to invoke err callback when before flush started to prevent user from sending more?

Reply:
if flush (initiated by disconnect_cb) is completed, it will be easier to move EP to err state, wdyt?

Reply:
maybe easier to code this, but it's not correct behavior
error handler should be called first, when we flush the lanes they must already be disabled.
maybe need to create wireup_ep which will wrap the real transport which we flush, and will prevent the user from issuing any send operations

Reply:
ok, will move back for now